packages/gatsby/src/schema/schema.js

```diff
@@ -0,0 +1,362 @@
+          schemaComposer.addSchemaMustHaveType(typeComposer)
+        }
+      })
+    } else {
+      schemaComposer.add(typeOrTypeDef)
+    }
+  })
+}
+
+const addSetFieldsOnGraphQLNodeTypeFields = ({
+  schemaComposer,
+  nodeStore,
+  parentSpan,
+}) =>
+  Promise.all(
+    Array.from(schemaComposer.values()).map(async tc => {
+      if (
+        tc instanceof schemaComposer.TypeComposer &&
+        tc.hasInterface(`Node`)
+      ) {
+        const typeName = tc.getTypeName()
+        const [fields] = await apiRunner(`setFieldsOnGraphQLNodeType`, {
+          type: {
+            name: typeName,
+            nodes: nodeStore.getNodesByType(typeName),
+          },
+          traceId: `initial-setFieldsOnGraphQLNodeType`,
+          parentSpan: parentSpan,
+        })
+        if (fields) {
+          // NOTE: `setFieldsOnGraphQLNodeType` only allows setting
+          // nested fields with a path as property name, i.e.
+          // `{ 'frontmatter.published': 'Boolean' }`, but not in the form
+          // `{ frontmatter: { published: 'Boolean' }}`
+          tc.addNestedFields(fields)
+        }
+      }
+    })
+  )
+
+const addThirdPartySchemas = ({
+  schemaComposer,
+  thirdPartySchemas,
+  parentSpan,
+}) => {
+  thirdPartySchemas.forEach(schema => {
+    const QueryTC = schemaComposer.TypeComposer.createTemp(
+      schema.getQueryType()
+    )
+    const fields = QueryTC.getFields()
+    schemaComposer.Query.addFields(fields)
+
+    // Explicitly add the third-party schema's types, so they can be targeted
+    // in `addResolvers` API.
+    const types = schema.getTypeMap()
+    Object.keys(types).forEach(typeName => {
+      const type = types[typeName]
+      if (
+        type !== schema.getQueryType() &&
+        !isSpecifiedScalarType(type) &&
+        !isIntrospectionType(type)
+      ) {
+        schemaComposer.add(type)
+      }
+    })
+  })
+}
+
+const addCustomResolveFunctions = async ({ schemaComposer, parentSpan }) => {
+  const intermediateSchema = schemaComposer.buildSchema()
+  const createResolvers = resolvers => {
+    Object.keys(resolvers).forEach(typeName => {
+      const fields = resolvers[typeName]
+      if (schemaComposer.has(typeName)) {
+        const tc = schemaComposer.getTC(typeName)
+        Object.keys(fields).forEach(fieldName => {
+          const fieldConfig = fields[fieldName]
+          if (tc.hasField(fieldName)) {
+            const originalTypeName = tc.getFieldType(fieldName).toString()
+            const fieldTypeName =
+              fieldConfig.type && fieldConfig.type.toString()
+            if (
+              !fieldTypeName ||
+              tc.getFieldType(fieldName) === fieldConfig.type.toString()
+            ) {
+              const newConfig = {}
+              if (fieldConfig.args) {
+                newConfig.args = fieldConfig.args
+              }
+              if (fieldConfig.resolve) {
+                newConfig.resolve = fieldConfig.resolve
```

Comment:
I think I had left a comment about this already, but can't find it atm -- It would be good to make the original resolver available on `info` or smth for easy wrapping